File: app/src/main/java/torille/fi/lurkforreddit/data/RedditDataSource.java
package torille.fi.lurkforreddit.data;

import android.support.annotation.NonNull;
import android.support.v4.util.Pair;

import java.util.List;

import io.reactivex.Observable;
import torille.fi.lurkforreddit.data.models.view.Comment;
import torille.fi.lurkforreddit.data.models.view.Post;
import torille.fi.lurkforreddit.data.models.view.PostAndComments;
import torille.fi.lurkforreddit.data.models.view.SearchResult;
import torille.fi.lurkforreddit.data.models.view.Subreddit;

/**
 * Main entry point for accessing data.
 */
public interface RedditDataSource {

    Observable<List<Subreddit>> getSubreddits();

    Observable<Pair<String, List<Post>>> getSubredditPosts(@NonNull String subredditUrl);

    Observable<Pair<String, List<Post>>> getMoreSubredditPosts(@NonNull String subredditUrl,
                                                               @NonNull String nextpageId);

    void refreshData();

    Observable<PostAndComments> getCommentsForPost(@NonNull String permaLinkUrl, boolean isSingleCommentThread);

    Observable<List<Comment>> getMoreCommentsForPostAt(@NonNull List<String> childCommentIds,
                                                       @NonNull String linkId,
                                                       int commentLevel);

    Observable<Pair<String, List<SearchResult>>> getSearchResults(@NonNull String query);

    Observable<Pair<String, List<SearchResult>>> getMoreSearchResults(@NonNull String query,
                                                                      @NonNull String after);
}

File: app/src/main/java/torille/fi/lurkforreddit/data/RedditDataSource.kt
package torille.fi.lurkforreddit.data

import io.reactivex.Flowable
import io.reactivex.Observable
import torille.fi.lurkforreddit.data.models.view.*

/**
 * Main entry point for accessing data.
 */
interface RedditDataSource {

    fun getSubreddits(): Flowable<List<Subreddit>>

    fun saveSubreddit(subreddit: Subreddit)

    fun saveSubreddits(subreddits: List<Subreddit>)

    fun getSubredditPosts(subredditUrl: String): Observable<kotlin.Pair<String, List<Post>>>

    fun getMoreSubredditPosts(subredditUrl: String,
                              nextpageId: String): Observable<kotlin.Pair<String, List<Post>>>

    fun refreshData()

    fun getCommentsForPost(permaLinkUrl: String, isSingleCommentThread: Boolean): Observable<PostAndComments>

    fun getMoreCommentsForPostAt(childCommentIds: List<String>,
                                 linkId: String,
                                 commentLevel: Int): Observable<List<Comment>>

    fun getSearchResults(query: String): Flowable<kotlin.Pair<String, List<SearchResult>>>

    fun getMoreSearchResults(query: String,
                             after: String): Flowable<kotlin.Pair<String, List<SearchResult>>>
}
